### ISSUE TITLE
Avoid false-sharing

### ISSUE THREAD

**Issue by bartoszmodelski** [ocaml/ocaml#11986]
False-sharing is pretty bad for performance of lock-free algorithms. There's many known results. This classic [paper](https://dl.acm.org/doi/10.5555/1295480.1295483) is a little old but for example: 

* 80% reduction in wall-clock time on linear regression claimed by [Intel's coobkook](https://www.intel.com/content/www/us/en/develop/documentation/vtune-cookbook/top/tuning-recipes/false-sharing.html). 
* In the case of multicore OCaml, @anmolsahoo25 found a 23% reduction in wall-clock time on a [benchmark](https://github.com/anmolsahoo25/aligned_int) with just two threads.

We've had some discussion on [lockfree](https://github.com/ocaml-multicore/lockfree/issues/19). There's probably at least 2 acceptable solutions there. The rest of this issue describes both. Let me know what you think, I'm happy to open a PR with either. 

# Expose memory-aligned allocation 

A somewhat principled C-style solution is to just call `posix_memalign` / `aligned_alloc` (C11) to get a block of memory, which occupies entire cache line(s). It makes it impossible for the cache line to be falsely-shared. It's also useful for doing math quickly (SSE, AVX). 

I claim we can do this relatively easily by sending these allocations into `large_allocate` with a special path that uses `aligned_alloc` (rather than `malloc`). ~~Current threshold for large alloc is 129, while cache line is 64 on most systems, so we're pretty close to using this infra anyway.~~ 

See this approach [here](https://github.com/bartoszmodelski/ocaml/pull/2/files#diff-b468e47c34ab18ea7b2cd8322cde20a2fa4d3e647422ee725f0fe7d84f768a32). 

# Add padding

A nice workaround is to just add padding. If block spans multiple cache lines and the middle is accessed then alignment doesn't matter. @polytypic implemented it in [multicore-magic](https://github.com/ocaml-multicore/multicore-magic/blob/d06dab4b3abf591eae4ae16e700051e9c3dd33ae/src/main/Multicore_magic.ml#L3). A disadvantage here is that we cannot pad area before the atomic (or it wouldn't be a valid `Atomic.t`), so we can still have accidental false-sharing if some other atomics in the program have not been padded. The positive part is that it's super simple. 

If that's the blessed way to go forward, I would propose to add that to `Obj` or/and create `Atomic.make_padded`. 

# Other ideas

@sadiqj mentioned making `Atomic.t` unboxed to let us pad records like in Java or Golang. It sounds reasonable but I do not have the expertise to create a POC. 

cc @kayceesrk

**Comment by stedolan** [comment]
> Current threshold for large alloc is 129, while cache line is 64 on most systems, so we're pretty close to using this infra anyway.

This isn't quite right: the threshold for large allocations is 129 *words*, which is 8256 bytes on 64-bit systems.

In fact, I think the slab allocator should already be cache-aligned for allocations of the right size. Perhaps we should make Atomic.t be a 7-word record, so that with the header it consumes 64 bytes on 64-bit systems?

**Comment by polytypic** [comment]
False sharing is mostly problematic with long lived objects.  Simply making all atomic objects cache aligned will not give best overall performance.  There are a lot of use cases of atomics, such as for promises or lazy values, where the atomic is only written a couple of times after initialization. Allocating more memory for those will likely degrade performance.

Also, it is important to understand that false sharing is not at all limited to atomic objects and does not even require any atomic objects.  You could, for example, have a regualar ref that is only used by the domain that allocated it for its own purposes, allocated between two read-only objects that are shared to other domains and you would get false sharing.

**Comment by polytypic** [comment]
If you look at the documentation of [`copy_as_padded`](https://ocaml-multicore.github.io/multicore-magic/multicore-magic/Multicore_magic/index.html#val-copy_as_padded) in the multicore-magic library, it says

> This is designed to help avoid [false sharing](https://en.wikipedia.org/wiki/False_sharing). False sharing has a negative impact on multicore performance. Accesses of both atomic and non-atomic locations, whether read-only or read-write, may suffer from false sharing.
>
> The intended use case for this is to pad all long lived objects that are being accessed highly frequently (read or written).

and then there are example of using `copy_as_padded` on many kinds of objects and that is very much intentional, because only padding atomic objects will not be enough.

Again, I want to remind that `copy_as_padded` is a workaround and not an ideal solution.  Something like `copy_as_cache_aligned` would be more effective and likely less wasteful.  Even that is not perfect.  You will sometimes want to e.g. have a record where certain fields should ideally be in different cache lines.

**Comment by stedolan** [comment]
The multicore allocator tries to allocate objects from domain-local pages. So you'll get false sharing if you allocate three objects sequentially and send the middle one to another domain while using the other two locally, but in general you should not get false sharing between objects allocated on one domain and objects allocated on another.

In other words, to avoid false sharing it is good to take a copy of small objects on the receiving domain, but I don't think you necessarily need to pad.

**Comment by polytypic** [comment]
Yes, with detailed knowledge of how the OCaml GC operates, it may be possible in some cases to design algorithms on top of it that avoid false sharing &mdash; at least when you consider such an algorithm in isolation.

I would caution against relying on such knowledge too much.  What happens, for example, if compaction is revived?  Assumptions about what objects could end up next to each other might change and one might then have to rewrite libraries and applications built on such intricate knowledge of the previous runtime.  Similar problems could happen with other changes to the GC.  If enough code is written with some specific assumptions about how the GC internally works, then that could impede future improvements to the GC, due to the cost in terms of having to rewrite substantial parts of the ecosystem.  (I guess that was already a thing with the transition to multicore.)

In other words, I'd suggest coming up with approaches that are stable with respect to potential future changes.  For example, being able to say that an object should be allocated in a cache line granular manner seems like such an approach.  No matter how the GC might change in future, as long as that one guarantee remains, one can use it to reliably avoid false sharing.

Likewise, I suggest coming up with approaches that are composable.  What I mean by that is than even if some algorithm could be devised to independently avoid false sharing (e.g. by allocating things on particular domains), then what guarantee there is that other code does not potentially interfere.  For example, other code, running in the same domain, could allocate refs, that might end up next to a supposedly false sharing free object, and share or mutate those refs causing false sharing.  That could even happen with the approach of copying objects on the receiving domain.  Perhaps there are ways to reliably and composably avoid false sharing by means other than having objects allocated in a cache line granular manner (padding is an approximation of that as workaround) and if there are, I'd love to see a detailed explanation of how a general multicore algorithm can be simply changed to take advantage of such an approach and an argument of why it prevents false sharing.

**Comment by bartoszmodelski** [comment]
> This isn't quite right: the threshold for large allocations is 129 _words_, which is 8256 bytes on 64-bit systems.

Rip me. Thanks for pointing out. 
 
> In fact, I think the slab allocator should already be cache-aligned for allocations of the right size. 

I assume that slabs refers pools in the heap code. As far as I can tell the memory for pools is cache-aligned, but the allocations themselves are not. I tried the following: 

```
value var = caml_alloc_shr(7, 0);
assert((var - 8) % 64 == 32); // succeeds 
```
Could be a red herring, but it coincides with POOL_HEADER_WSIZE = 4.

> Perhaps we should make Atomic.t be a 7-word record, so that with the header it consumes 64 bytes on 64-bit systems?

I assume that 8-word pool will only keep 8-word chunks even under sweeping or compaction. With pool-allocs being aligned, this proposition sounds okay. All the assumptions here seem foundational. Ideally, maybe we can wrap it into a separate call? It will make it easier for the users, and easier for the compiler folks to evolve if needed. 

However, I'm not sure if making all Atomic.t 7-word is optimal. It's a better world than now, but I would not want to lose current atomic. To add to the list, I can see myself using both (padded and thin atomic) even within the same data structure. For example, a queue with contention in FAD-manipulated indices but also requiring cells to be atomic for correctness.

**Comment by polytypic** [comment]
Speaking of the "fat atomic" proposal (i.e. make the size of `Atomic.t` a total of 8 words).  I basically tried that in one of my experiments in `par-ml`.  I used 16 words for an atomic as cache line size is 128 bytes on Apple M1.

I basically replaced the `Stdlib.Atomic` module with my own (this is typed here from memory forillustration only):

```ocaml
module Atomic = struct
  include Stdlib.Atomic
  type 'a padded = {
    value: 'a;
    p2 : int;
    p3 : int;
    p4 : int;
    p5 : int;
    p6 : int;
    p7 : int;
    (* I used up to 15 *)
  }
  let make value = Obj.magic {value; p2=0; p3=0; p4=0; p5=0; p6=0; p7=0}
end
```

The end result was that the program ran significantly slower.  I would assume part of the reason for that was that only a few of the atomics used in the program suffered from false-sharing.  The other part is the additional writes to initialize the padding words and the extra cache pressure caused by higher memory usage.  The vast majority of atomics were allocated for only short term use and were accessed a couple of times as they essentially represented promise like cells.

So, I would not recommend the approach of making all atomics 8 (or 16) words.  It will likely not improve performance in general, because making atomics 8 words (or 16 words) will increase the amount of cache lines required by a factor of 4 (or 8) and the cost of that is likely greater than the rare win from avoiding some accidental false-sharing.

**Comment by Octachron** [comment]
cc @sadiqj would you have the time to have a look at the issue ? I am temporarily assigning you (with the semantics that assignee people should try to shepherd discussions about issues but not necessarily fix them), but please free to de-assign yourself (or even better to assign someone else).

**Comment by bartoszmodelski** [comment]
I’m happy to shoot a PR following @stedolan’s suggestion in a week or two
(on holidays now).

**Comment by gasche** [comment]
I think that this can be closed following #12212, although of course there are other sources of false sharing than atomic values that could degrade program performance.
